### PR TITLE
[MRG] Addition of Simplex SVR model and little fix on SVR

### PR DESCRIPTION
**Pull request by Klopfe**
Tests for SVR are still pretty slow and this is probably due to the choice of stopping criterion based on tol * pobj0 when pobj0 is the initial primal objective. 
However, we are solving in the dual and in the tests for example it leads to tol * pobj0 to be smaller than 1e-18 and the dual gap never reaches this tolerance so it goes up to max iter. 

I tried to lower the tol but then enet tests break.

All tests are clear though.

### CODE REVIEW

File: doc/api.rst
Comment (mathurinm) [inline doc/api.rst:32]: nitpick: do we put them in alphabetical order ?

File: sparse_ho/tests/common.py
Comment (mathurinm) [inline sparse_ho/tests/common.py:113]: isn't np.log of np.geomspace just a np.linspace ?

**Comment by mathurinm** [comment]
Thanks a lot @Klopfe ! I didn't read the diff for `ssvr.py` since it's quite large, do you think there is a way to factorize a bit with `svr.py` ?

When you say "I tried to lower the tol but then enet tests break." you mean to increase the tol ? Indeed 1e-15 is super low. Maybe we could increase it, and increase `atol` and `rtol` in the assertions ?

File: doc/api.rst
Comment (Klopfe) [reply doc/api.rst]: you're right. It's been changed

**Comment by Klopfe** [comment]
Thanks for looking at it. Yes, I think there would be ways to factorize. 
I will try to do that next.

**Comment by Klopfe** [comment]
It's a proposition but SSVR could be a child (in the object programming sense) of SVR so that there would be a decent amount of functions that would not need to be rewritten. 
What do you think?

**Comment by mathurinm** [comment]
Yes, that makes a lot of sense! I don't know how the two classes differ, otherwise you may want a `BaseSVR` class that both `SVR` and `SimplexSVR` inherit from. As you see fit!

**Comment by Klopfe** [comment]
yes I will see what's best. Thanks!

**Comment by mathurinm** [comment]
On master tests run in 50 s, on this PR they take 134 s, I think we cannot merge as is. Do you have the same @Klopfe ?

This branch:
```
================== 1 failed, 73 passed, 10 xfailed, 748 warnings in 134.08s (0:02:14) ==================

========================================= slowest 10 durations =========================================
46.59s call     sparse_ho/tests/test_models.py::test_check_grad_sparse_ho[algo1-svr-MSE]
28.29s call     sparse_ho/tests/test_models.py::test_check_grad_sparse_ho[algo1-ssvr-MSE]
12.17s call     sparse_ho/tests/test_models.py::test_beta_jac[ssvr]
5.88s call     sparse_ho/tests/test_models.py::test_beta_jac[svr]
3.85s call     sparse_ho/tests/test_models.py::test_check_grad_sparse_ho[algo1-enet-MSE]
3.62s call     sparse_ho/tests/test_models.py::test_beta_jac[enet]
3.42s call     sparse_ho/tests/test_models.py::test_beta_jac[wLasso]
3.16s call     sparse_ho/tests/test_models.py::test_check_grad_sparse_ho[algo1-svm-MSE]
2.80s call     sparse_ho/tests/test_models.py::test_beta_jac[logreg]
1.95s call     sparse_ho/tests/test_criterion.py::test_cross_val_criterion[wLasso-criterion2]
```

master:
```
=========== 1 failed, 60 passed, 15 xfailed, 515 warnings in 50.92s ============
```

**Comment by Klopfe** [comment]
Yeah, it runs in 100 seconds here. So it doubles the time. I will try to find a fix. I can play with the regularization hyperparameter and maybe lower the tol.

**Comment by mathurinm** [comment]
but is there a fundamental reason for it to much slower than test for existing estimators ?

**Comment by Klopfe** [comment]
well, the optimization problem is two times bigger than the svm ones. But it shouldn't be that long for our tests.

**Comment by mathurinm** [comment]
It may be interesting to isolate the slow test and run it in max verbosity mode to see visually what takes time

**Comment by Klopfe** [comment]
yes, I think I caught a bigger problem than just svr. Hard to explain typing though... 
It's related to warm start in test and the fact that the Jacobian does not need to be recomputed sometimes. 
When that happens the stopping criterion to stop the Jacobian calculation is never stopped and goes to max_iter_jac.

I push a temporary commit print the problem. To see it just execute test_models in ipython. 
Here its for the Lasso (to show that it's not only SVR related)

I commented lines 154-155 of implicit_forward.py a temporary fix but we might need a bigger fix in the warm start section.

**Comment by mathurinm** [comment]
nice catch ! since it's unrelated to SVR, any idea why this did not pop up before ?

Is it ok for you to tackle that in another PR first ? From a design point of view I don't understand why we enter into the recomputation of the Jacobian if we can use the old one, what am I missing ?

**Comment by Klopfe** [comment]
yes I can tackle that in another PR. I have to understand why we rerun two times the same optimization problem with the same data and parameters and maybe add a check somewhere. 

I will clean the files for the test and try to work on the factorization in the next few days on this PR.

**Comment by mathurinm** [comment]
just to be clear, I think it would be better to address the warm start issue before finishing this PR, so that we don't merge tests that are too long. My 2 cents, your call in the end

**Comment by Klopfe** [comment]
agreed

**Comment by mathurinm** [comment]
After merging master,I still have some slow tests for ssvr, with algo0 and algo2 . Could it be related to the same bug for other algos ?

```
================================================== slowest 20 durations ==================================================
41.60s call     sparse_ho/tests/test_models.py::test_check_grad_sparse_ho[algo0-ssvr-MSE]
23.79s call     sparse_ho/tests/test_models.py::test_check_grad_sparse_ho[algo2-ssvr-MSE]
6.67s call     sparse_ho/tests/test_models.py::test_beta_jac[ssvr]
5.15s call     sparse_ho/tests/test_models.py::test_beta_jac[svr]
3.31s call     sparse_ho/tests/test_models.py::test_beta_jac[wLasso]
2.80s call     sparse_ho/tests/test_models.py::test_beta_jac[enet]
```

**Comment by Klopfe** [comment]
I will look into it.

**Comment by Klopfe** [comment]
@mathurinm can you check that the tests run fine now for you ? 
There was a weird bug where due to the choice of y to be -1 or 1, p_obj0 was 0 and we went up to max_iter again in the iterations.

**Comment by mathurinm** [comment]
40 s on my machine :1st_place_medal: 

I don't know why, two unrelated tests are very slow on github actions.

**Comment by Klopfe** [comment]
Yes I saw that too. More than 30second each. 
I don't know what affects these two tests. It is with the line-search

**Review by mathurinm — APPROVED** [review]

**Pull request merged**
